sae/blocks.go

```diff
@@ -104,74 +101,148 @@ func (vm *VM) buildBlock(
+					zap.Int("tx_index", i),
+					zap.Stringer("tx_hash", tx.Hash()),
+					zap.Error(err),
+				)
+				return nil, fmt.Errorf("applying tx %#x in block %d to worst-case state: %v", tx.Hash(), b.Height(), err)
+			}
 		}
-		if err := clock.AfterBlock(lim, vm.hooks, b.Header()); err != nil {
-			return nil, err
+		for i, op := range vm.hooks.EndOfBlockOps(b.EthBlock()) {
+			if err := state.Apply(op); err != nil {
+				log.Warn("Could not apply op during historical worst case calculation",
+					zap.Int("op_index", i),
+					zap.Stringer("op_id", op.ID),
+					zap.Error(err),
+				)
+				return nil, fmt.Errorf("applying op at end of block %d to worst-case state: %v", b.Height(), err)
+			}
+		}
+		if err := state.FinishBlock(); err != nil {
+			log.Warn("Could not finish historical worst case calculation",
+				zap.Error(err),
+			)
+			return nil, fmt.Errorf("finishing worst-case state for block %d: %v", b.Height(), err)
 		}
 	}
 
+	// TODO: This header hasn't had an opportunity to be populated by,
+	// [hook.Points] but it is passed into [hook.Points.SubSecondBlockTime]. We
+	// should construct the header inside of [hook.Points].
 	hdr := &types.Header{
 		ParentHash: parent.Hash(),
-		Number:     new(big.Int).Add(parent.Number(), common.Big1),
-		Time:       unix(blockTime),
 		Root:       lastSettled.PostExecutionStateRoot(),
+		Number:     new(big.Int).Add(parent.Number(), common.Big1),
+		// GasLimit is populated after [worstcase.State.StartBlock].
+		// GasUsed should be populated after [worstcase.State.ApplyTx] has
+		// finished being called.
+		Time: unix(blockTime),
+		// BaseFee is populated after [worstcase.State.StartBlock].
+
+		// All other fields are, optionally, populated by the call to
+		// [hook.BlockBuilder.BuildBlock] after applying every
+		// [types.Transaction].
+	}
+	if err := state.StartBlock(hdr); err != nil {
+		log.Warn("Could not start worst case block calculation",
+			zap.Error(err),
+		)
+		return nil, fmt.Errorf("starting worst-case state for new block: %v", err)
 	}
-	clock.BeforeBlock(vm.hooks, hdr)
-	hdr.BaseFee = clock.BaseFee().ToBig()
 
-	var included []*types.Transaction
+	hdr.GasLimit = state.GasLimit()
+	hdr.BaseFee = state.BaseFee().ToBig()
+
+	var (
+		candidates = pendingTxs(txpool.PendingFilter{
+			BaseFee: state.BaseFee(),
+		})
+		included []*types.Transaction
+	)
 	for _, ltx := range candidates {
+		// If we don't have enough gas remaining in the block for the minimum
+		// gas amount, we are done including transactions.
+		if remainingGas := state.GasLimit() - state.GasUsed(); remainingGas < params.TxGas {
+			break
+		}
+
 		tx, ok := ltx.Resolve()
 		if !ok {
 			continue
 		}
-		_ = sdb // balance checks
+
+		if err := state.ApplyTx(tx); err != nil {
+			continue
+		}
 		included = append(included, tx)
 	}
 
+	// TODO: Should the [hook.BlockBuilder] populate [types.Header.GasUsed] so
+	// that [hook.Op.Gas] can be included?
```

Comment:
Please explain. How is `hook.Op.Gas` not included if you're using `state.GasUsed()`, which has had the ops applied?